### PR TITLE
Address mina-signer audit, Pt 2: Low-level crypto tests

### PR DESCRIPTION
**Pull request by mitschabaude**
Successor to https://github.com/o1-labs/snarkyjs/pull/714

Addresses the following parts of the audit -- tests for low-level crypto arithmetic:

> ## 2 Test coverage
> Result: FAIL
> * js_crypto's field arithmetic methods do not have individual unit tests
> * js_crypto should have low-level tests to verify fundamental group operations
>   * Addition is commutative X + Y == Y + X
>   * Associativity X + (Y + Z) == (X + Y) z
>   * Identity I + X == X + I == X
>   * Inverse X + inv(X) == inv(X) + X == I
>   * Scaling commutes with adding scalars A * (x + y) == A * x + A * y
>   * Scaling commutes with multiplying scalars A * (x * y) == (x * y) * A
>   * Scaling commites with negation A * neg(x) == neg(A * x)
>   * Always test on curve after scaling
>   * Note: These are easy to add, so highly recommended to implement
> * js_crypto primitives sub, inverse, div, square, fromNumber and fromBigint do not have any test coverage

This PR ended up making some changes to the curve arithmetic, so in order to enable a focused review I'm leaving other missing tests for a third PR.

* Added tests for finite field and curve arithmetic
* Fixed issues that were uncovered by these tests:
  * added missing methods `Group.equal`, `Group.isOnCurve` to be able to test
  * `Group.add` was not complete addition -- failed for `Group.add(X, X)`. I added the necessary logic to fall back to doubling. Also handles the `Group.add(X, -X) = 0` case now to return zero in the default representation
* For completeness, added implementation and tests for endomorphism

### CODE REVIEW

File: src/lib/account_update.ts [hunk 1]
```diff
@@ -1877,7 +1877,11 @@ function signJsonTransaction(
       accountUpdate.authorization.proof === null
     ) {
       zkappCommand = JSON.parse(
-        Ledger.signAccountUpdate(JSON.stringify(zkappCommand), privateKey, i)
+        Ledger.signOtherAccountUpdate(
+          JSON.stringify(zkappCommand),
```
Comment (jspada) [inline src/lib/account_update.ts:1881]: Do you need to worry about `stringify` safety wrt circular references?

File: src/lib/testing/random.ts [hunk 2]
```diff
@@ -0,0 +1,501 @@
...
+      };
+    },
+  };
+}
+
+/**
+ * uniform distribution over range [min, max]
+ * with bias towards special values 0, 1, -1, 2, min, max
+ */
+function int(min: number, max: number): Random<number> {
```
Comment (jspada) [inline src/lib/testing/random.ts:351]: cool :-)

File: src/mina-signer/tests/verify-in-snark.unit-test.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      },
+    },
+  },
+});
+
+await MyProgram.compile();
+let proof = await MyProgram.verifySignature(null, signature);
+ok = await MyProgram.verify(proof);
+expect(ok).toEqual(true);
+
```
Comment (jspada) [inline src/mina-signer/tests/verify-in-snark.unit-test.ts:48]: How about some negative tests?

File: src/lib/account_update.ts (hunk 1 above)
Comment (mitschabaude) [reply src/lib/account_update.ts]: nope, none of these transaction data structures has any circular references (and they are kept really explicitly typed so something like an accidental circular ref in there shouldn't happen)

File: src/mina-signer/tests/verify-in-snark.unit-test.ts (hunk 3 above)
Comment (mitschabaude) [reply src/mina-signer/tests/verify-in-snark.unit-test.ts]: good idea!

**Pull request merged**